### PR TITLE
JAMES-3775 Mailet for RSpamD

### PR DESCRIPTION
**Pull request by vttranlina**

### CODE REVIEW

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/RSpamDScanner.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+import org.apache.mailet.Mail;
+import org.apache.mailet.PerRecipientHeaders;
+import org.apache.mailet.base.GenericMailet;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableList;
+
+public class RSpamDScanner extends GenericMailet {
+    public static final AttributeName FLAG_MAIL = AttributeName.of("org.apache.james.spamassassin.flag");
```
Comment (vttranlina) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/RSpamDScanner.java:47]: Should we still keep using this attribute?
Or another for less confuse 
eg: `org.apache.james.rspamd.flag`
But if we use another attribute, we need to modify the `IsMarkedAsSpam` matcher

Comment (Arsnael) [reply server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/RSpamDScanner.java:47]: I would agree on something more generic if we have more than one spam implem yes
like `org.apache.james.spam.flag` maybe?

**Comment by Arsnael** [comment]
Read it

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/RSpamDScanner.java (hunk 1 above)
Comment (vttranlina) [reply server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/RSpamDScanner.java:47]: I thought again, 
The attribute should be unique, eg case we use multi spam 3rd (both rspamd and spamassian)
IMO, we will keep `org.apache.james.spamassassin.x` for spamassian,
`org.apache.james.rspamd.x` for rspamd. 
And modify the `IsMarkedAsSpam` for reading attributeName from mailet XML configuration file.

Comment (Arsnael) [reply server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/RSpamDScanner.java:47]: Ok to have distinct flags
However let's be honest, there is nothing logical for a environment to use 2 anti spam third party softwares at the same time :)

File: server/mailet/mailets/pom.xml [hunk 2]
```diff
@@ -171,6 +177,12 @@
             <artifactId>james-server-queue-memory</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>${james.groupId}</groupId>
+            <artifactId>james-server-rate-limiter-redis</artifactId>
+            <type>test-jar</type>
+            <scope>test</scope>
+        </dependency>
```
[the remark this answers is not in the document]
Comment (vttranlina) [reply server/mailet/mailets/pom.xml:185]: for RSpamDExtension, when without Redis dependency, it throws a RedisExtension not found

File: server/mailet/mailets/src/test/java/org/apache/james/transport/matchers/IsMarkedAsSpamTest.java [hunk 3]
```diff
@@ -151,4 +152,29 @@ void isMarkedAsSpamShouldMatchWhenHeaderAndYesValueInOtherCase() throws Exceptio
...
+        FakeMail mail = FakeMail.builder()
+            .name("name")
+            .sender("sender@james.org")
+            .recipient("to@james.org")
+            .addHeaderForRecipient(PerRecipientHeaders.Header.builder()
+                    .name("custom.package")
+                    .value("Yes, hits=6.8 required=5.0")
+                    .build(),
+                new MailAddress("to@james.org"))
+            .attribute(Attribute.convertToAttribute("custom.package", "Yes, hits=6.8 required=5.0"))
```
[the remark this answers is not in the document]
Comment (vttranlina) [reply server/mailet/mailets/src/test/java/org/apache/james/transport/matchers/IsMarkedAsSpamTest.java:174]: Can you explain more?
I don't get the point

[the remark this answers is not in the document]
Comment (quantranhong1999) [reply server/mailet/mailets/src/test/java/org/apache/james/transport/matchers/IsMarkedAsSpamTest.java:174]: E.g:
```java
.attribute(Attribute.convertToAttribute("custom.package", "Yes, actions=reject score=14.225 requiredScore=14.0 desiredRewriteSubject="))
```

**Comment by quantranhong1999** [comment]
```shell
Failed
org.apache.james.transport.mailets.RSpamDScannerTest.serviceShouldWriteMessageAsNotSpamWhenNotSpam

expected: PerRecipientHeaders{headersByRecipient={user1@exemple.com=[Header{name=org.apache.james.rspamd.flag, value=NO}, Header{name=org.apache.james.rspamd.status, value=No, actions=no action score=2.9 requiredScore=14.0 desiredRewriteSubject=}]}}
 but was: PerRecipientHeaders{headersByRecipient={user1@exemple.com=[Header{name=org.apache.james.rspamd.flag, value=NO}, Header{name=org.apache.james.rspamd.status, value=No, actions=no action score=2.4 requiredScore=15.0 desiredRewriteSubject=}]}}
```

Likely need to put the rspamd conf file here. BTW the same scoring issue.

**Review by Arsnael — APPROVED** [review]

File: server/apps/distributed-app/docs/modules/ROOT/partials/IsMarkedAsSpam.adoc [hunk 4]
```diff
@@ -14,4 +14,15 @@ As an example, here is a part of a mailet pipeline which can be used in your Loc
...
+
+In order to use this with `rspamd`, we need to declare a condition for the matcher.
+Eg: With the recipient header for RSpamD being *org.apache.james.rspamd.status*,
+then the configuration would be:
+
+....
+<!-- SpamAssassing mailets pipeline -->
+    <mailet match="IsMarkedAsSpam=org.apache.james.rspamd.status" class="WithStorageDirective">
+        <targetFolderName>Spam</targetFolderName>
+    </mailet>
```
Comment (chibenwa) [inline server/apps/distributed-app/docs/modules/ROOT/partials/IsMarkedAsSpam.adoc:27]: <3

File: server/mailet/mailets/pom.xml [hunk 5]
```diff
@@ -63,6 +63,16 @@
             <artifactId>apache-james-mailbox-store</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>${james.groupId}</groupId>
+            <artifactId>apache-james-rspamd</artifactId>
```
Comment (chibenwa) [inline server/mailet/mailets/pom.xml:68]: Now we do not want the RSPAMD dependency here!

RSPAMD stuff should be in `third-party/rspamd` package.

Please relocate/cleanup everything!

File: server/mailet/mailets/pom.xml [hunk 6]
```diff
@@ -161,6 +177,12 @@
             <artifactId>james-server-queue-memory</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>${james.groupId}</groupId>
+            <artifactId>james-server-rate-limiter-redis</artifactId>
+            <type>test-jar</type>
+            <scope>test</scope>
+        </dependency>
```
Comment (chibenwa) [inline server/mailet/mailets/pom.xml:185]: Why?

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/RSpamDScanner.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    static final String RSPAMD_TIMEOUT = "rSpamDTimeout";
+
+    private RSpamDHttpClient rSpamDHttpClient;
+
+    @Override
+    public void init() throws MessagingException {
+        Optional<Integer> rSpamDTimeoutConfigure = Optional.ofNullable(getInitParameter(RSPAMD_TIMEOUT))
+            .filter(s -> !s.isEmpty())
+            .map(Integer::valueOf)
+            .filter(i -> i > 0);
```
Comment (chibenwa) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/RSpamDScanner.java:60]: Instead we should explicitly fail when the value is invalid eg `-1`!

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/RSpamDScanner.java [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+
+        RSpamDClientConfiguration rSpamDClientConfiguration = new RSpamDClientConfiguration(extractRSpamDUrl(),
+            getInitParameter(RSPAMD_PASSWORD, ""), rSpamDTimeoutConfigure);
+        rSpamDHttpClient = new RSpamDHttpClient(rSpamDClientConfiguration);
```
Comment (chibenwa) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/RSpamDScanner.java:64]: Question: should this be injected?

https://github.com/apache/james-project/pull/1092 injects this, we likely should be consistant...

File: server/mailet/mailets/src/main/java/org/apache/james/transport/matchers/IsMarkedAsSpam.java [hunk 9]
```diff
@@ -51,9 +54,18 @@ public class IsMarkedAsSpam extends GenericMatcher {
...
-        return "Has org.apache.james.spamassassin.status per recipient header with a Yes value";
+        return "Has " + spamStatusHeader + " per recipient header with a Yes value";
+    }
+
+    @Override
+    public void init() throws MessagingException {
+        spamStatusHeader = Optional.ofNullable(getCondition())
+            .filter(s -> !s.isEmpty())
+            .orElse(SpamAssassinResult.STATUS_MAIL.asString());
     }
```
Comment (chibenwa) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/matchers/IsMarkedAsSpam.java:69]: Document this change (javadoc above, in `server/apps/distributed-app/docs/modules/ROOT/partials/IsMarkedAsSpam.adoc`)

File: server/mailet/mailets/src/test/resources/eml/spam8.eml [hunk 10]
```diff
@@ -0,0 +1,321 @@
+From ormlh@imail.ru  Sun Jul 15 04:56:31 2001
```
Comment (chibenwa) [inline server/mailet/mailets/src/test/resources/eml/spam8.eml:1]: Quick question: where is this mail taken from? Protected by some sort of license?

Can we spend 5 minutes checking there is no issue?

File: server/mailet/mailets/src/test/resources/rspamd-config/actions.conf [hunk 11]
```diff
@@ -0,0 +1,29 @@
+# Actions settings
```
Comment (chibenwa) [inline server/mailet/mailets/src/test/resources/rspamd-config/actions.conf:1]: Why is this file needed?

File: server/mailet/mailets/pom.xml (hunk 6 above)
Comment (vttranlina) [reply server/mailet/mailets/pom.xml:185]: https://github.com/apache/james-project/pull/1093#discussion_r931785014

File: server/mailet/mailets/src/test/resources/eml/spam8.eml (hunk 10 above)
Comment (vttranlina) [reply server/mailet/mailets/src/test/resources/eml/spam8.eml:1]: > Where?

I cloned it from `third-party/spamassassin/src/test/resources/spamassassin_db/spam/spam8`

File: server/mailet/mailets/src/test/resources/rspamd-config/actions.conf (hunk 11 above)
Comment (vttranlina) [reply server/mailet/mailets/src/test/resources/rspamd-config/actions.conf:1]: The default rejects score of rspamd is 15. 
It will take more time if try to find the mail sample that has a score > 15. 
We already have a mail sample that has scored is 14.25, so it is easier to custom rspamd to decrease the threshold of action


More: 
https://github.com/apache/james-project/pull/1088#issuecomment-1195294537
https://github.com/apache/james-project/pull/1093#issuecomment-1198874573

**Comment by chibenwa** [comment]
Is `desiredRewriteSubject` supported?

File: server/mailet/mailets/src/test/resources/eml/spam8.eml (hunk 10 above)
Comment (chibenwa) [reply server/mailet/mailets/src/test/resources/eml/spam8.eml:1]: Ok but had this been decently checked upon that time? My comment is still relevant.

File: server/mailet/mailets/pom.xml (hunk 6 above)
Comment (chibenwa) [reply server/mailet/mailets/pom.xml:185]: Can you write a plain english response please? (link not working here.)

File: server/mailet/mailets/src/test/resources/eml/spam8.eml (hunk 10 above)
Comment (vttranlina) [reply server/mailet/mailets/src/test/resources/eml/spam8.eml:1]: checked
It's a unavailable email

File: server/mailet/mailets/pom.xml (hunk 6 above)
Comment (vttranlina) [reply server/mailet/mailets/pom.xml:185]: > for RSpamDExtension, when without Redis dependency, it throws a RedisExtension not found

Comment (vttranlina) [reply server/mailet/mailets/pom.xml:185]: Finally, it will be removed, because this mailet will be move  to third-party/rspamd

**Comment by vttranlina** [comment]
> Is `desiredRewriteSubject` supported?

it is already supported by rspamD, but it looks unnecessary for mailet, The mailet just cares YES/NO result for marking spam status.

**Comment by chibenwa** [comment]
IMO it could be interesting having the RSpamD mailet modifying the subject if need be. 

We can handle that in another issue.

File: server/mailet/mailets/src/main/java/org/apache/james/transport/matchers/IsMarkedAsSpam.java (hunk 9 above)
Comment (vttranlina) [reply server/mailet/mailets/src/main/java/org/apache/james/transport/matchers/IsMarkedAsSpam.java:69]: Your mean is do not use `SpamAssassinResult.STATUS_MAIL` by default?

IMO, we should keep it, it won't make degrade the old configure when not provided status package

File: server/apps/distributed-app/docs/modules/ROOT/partials/IsMarkedAsSpam.adoc [hunk 12]
```diff
@@ -14,4 +14,15 @@ As an example, here is a part of a mailet pipeline which can be used in your Loc
...
     </mailet>
 <!-- End of SpamAssassing mailets pipeline -->
+....
+
+In order to use this with `rspamd`, we need to declare a condition for the matcher.
+Eg: With the recipient header for RSpamD being *org.apache.james.rspamd.status*,
+then the configuration would be:
+
+....
+<!-- SpamAssassing mailets pipeline -->
```
Comment (chibenwa) [inline server/apps/distributed-app/docs/modules/ROOT/partials/IsMarkedAsSpam.adoc:24]:
```suggestion
<!-- RSpamD mailets pipeline -->
```

You should put the RSpamD mailet prior this. Also you need to explain that one would need to drop the RSpamD jar in `extensions-jars` folder.

File: third-party/rspamd/pom.xml [hunk 13]
```diff
@@ -18,9 +17,7 @@
     specific language governing permissions and limitations
     under the License.
 -->
-<project xmlns="http://maven.apache.org/POM/4.0.0"
-         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
```
Comment (chibenwa) [inline third-party/rspamd/pom.xml:20]: This change is not needed

File: third-party/rspamd/src/main/java/org/apache/james/rspamd/RSpamDScanner.java [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+
+    @Override
+    public void service(Mail mail) throws MessagingException {
+        MimeMessage mimeMessage = mail.getMessage();
+        AnalysisResult rSpamDResult;
+        try {
+            rSpamDResult = rSpamDHttpClient.checkV2(mimeMessage.getInputStream()).block();
+        } catch (IOException e) {
+            throw new RuntimeException("Error when RSpamD scan the message", e);
+        }
```
Comment (chibenwa) [inline third-party/rspamd/src/main/java/org/apache/james/rspamd/RSpamDScanner.java:78]: Extract that to a method in order to avoid variable reallocation please

File: third-party/rspamd/src/main/java/org/apache/james/rspamd/RSpamDScanner.java [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+            getInitParameter(RSPAMD_PASSWORD, ""), rSpamDTimeoutConfigure);
+        rSpamDHttpClient = new RSpamDHttpClient(rSpamDClientConfiguration);
+    }
+
+    @Override
+    public void service(Mail mail) throws MessagingException {
+        MimeMessage mimeMessage = mail.getMessage();
+        AnalysisResult rSpamDResult;
+        try {
+            rSpamDResult = rSpamDHttpClient.checkV2(mimeMessage.getInputStream()).block();
```
Comment (chibenwa) [inline third-party/rspamd/src/main/java/org/apache/james/rspamd/RSpamDScanner.java:75]: Use `new MimeMessageInputStream(mimeMessage)` instead.

**Review by chibenwa — APPROVED** [review]
Looks overall good to me.

**Comment by vttranlina** [comment]
Rebase master & squash fixup

**Review by Arsnael — APPROVED** [review]

**Review by quantranhong1999 — APPROVED** [review]

**Comment by quantranhong1999** [comment]
`serviceShouldWriteMessageAsNotSpamWhenNotSpam` is failing on CI and `serviceShouldWriteMessageAsSpamWhenSpam` is failing my local (score issue again). 
IMO we should change the assertion way to ignore the score.

**Comment by chibenwa** [comment]
Can you rebase?
